### PR TITLE
magento/magento2#22856: Catalog price rules are not working with custom options as expected.

### PR DESCRIPTION
**Pull request by p-bystritsky**
### Fixed Issues (if relevant)

1. magento/magento2#22856: Catalog price rules are not working with custom options as expected.

### Manual testing scenarios (*)

1. Create Product1 with fixed price custom options.
2. Create Product2 with percent price custom options.
3. Create and apply CatalogPriceRule1 with percent type discount.
4. Open Product1 and Product2 pages on the storefront and make sure discount is correctly applied to products options.
5. Disable CatalogPriceRule1.
6. Create and apply CatalogPriceRule2 with fixed type discount.
7. Open Product1 and Product2 pages on the storefront and make sure discount is correctly applied to products options.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Catalog/Model/Product/Option/Value.php [hunk 1]
```diff
@@ -225,16 +254,17 @@ public function saveValues()
     }
 
     /**
-     * Return price. If $flag is true and price is percent
-     *  return converted percent to price
+     * Return price.
+     *
+     * If $flag is true and price is percent return converted percent to price
      *
```
Comment (aleron75) [inline app/code/Magento/Catalog/Model/Product/Option/Value.php:262]: I take this comment as an example of very good and less good use of it at the same time. 

When you say something like "if $flag is true..." it is very valuable but at the same time saying "Return price" on a method named `getPrice()` doesn't add any additional value.

Next time, I would recommend avoiding comments that repeat the name of the method, just use a comment to explain something that's not very explicit, that way you will add valuable info.

File: app/code/Magento/Catalog/Pricing/Price/CustomOptionCatalogRulePriceCalculator.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        $this->priceCurrency = $priceCurrency;
+    }
+
+    /**
+     * Calculates prices of custom options of the product with catalog rules applied.
+     *
+     * @param ProductOptionValue $optionValue
+     * @return float
+     */
+    public function execute(
```
Comment (aleron75) [inline app/code/Magento/Catalog/Pricing/Price/CustomOptionCatalogRulePriceCalculator.php]: The `execute()` method underscores the functional usage of a service class, that's a correct approach. 

For this reason, a better name for the class would be one that suggests the command that is executed, like `CalculateCustomOptionCatalogRule`.

**Review by aleron75 — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Model/Product/Option/Value.php (hunk 1 above)
Comment (p-bystritsky) [reply app/code/Magento/Catalog/Model/Product/Option/Value.php:262]: @aleron75 thank you for the review! Unfortunately, code sniffers are forcing us to add some description to method annotation, so sometimes we have no other way then add some absolutely useless comment to old methods.

File: app/code/Magento/Catalog/Pricing/Price/CustomOptionCatalogRulePriceCalculator.php (hunk 2 above)
Comment (p-bystritsky) [reply app/code/Magento/Catalog/Pricing/Price/CustomOptionCatalogRulePriceCalculator.php]: @aleron75 done.

**Review by aleron75 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @aleron75, thank you for the review.
[ENGCOM-5122](https://jira.corp.magento.com/browse/ENGCOM-5122) has been created to process this Pull Request
:eight_spoked_asterisk: @aleron75, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by engcom-Delta** [comment]
Hi @p-bystritsky issue is also reproducible with custom option type Field. Could you take a look?
**Manual testing scenario:**
1. Create Product1 with Price=100$ and with fixed price custom options.
 - Option Type=Field
 - - Price=10$
 - - Price Type=Fixed
2. Create and apply CatalogPriceRule1 with percent type discount.
- Conditions:
- - Category  is {your category with products}
- Actions:
- - Apply=Apply as percentage of original
- - Discount Amount =50 
3. Open Product1 page on the storefront

**Result:**
Catalog rule discount  is applied only for base product price so final price is {{` (Product Price * Discount Amount) + Custom Option Price` }}
#22917Pr

**Comment by sidolov** [comment]
@p-bystritsky will you continue progress on PR?

**Comment by tkacheva** [comment]
@p-bystritsky - do you plan to merge this issue fix before 2.3.3 code freeze?

**Comment by p-bystritsky** [comment]
@sidolov, I'd like to continue.
@tkacheva, unfortunately, I'm not sure.

**Comment by engcom-Foxtrot** [comment]
@tkacheva, I need advice.
Lets' assume we have a Product with a price $100, Custom Option for that Product with price 10% and the active Catalog Price Rule with $50 discount. What should be the final price for that Product? $55 or $60?

**Comment by p-bystritsky** [comment]
@tkacheva, I need a consultation. It's not obvious how the Catalog Price Rule discount should be calculated in some cases.
Lets' assume we have a Product with a price $100, Custom Option for that Product with price 10% and the active Catalog Price Rule with $50 discount. What should be the final price for that Product? $55 or $60?

**Comment by johncarlson21** [comment]
where are we at with this? is this going to be fixed for 2.3.3? I am in dire need of this fix as our site relies on this.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by johncarlson21** [comment]
Hey guys.. just wondering where this is at and when this will be ready.

**Comment by johncarlson21** [comment]
@p-bystritsky  @aleron75 Hey guys.. where we at on this? will this be pushed to the next release?

**Comment by aleron75** [comment]
> @p-bystritsky @aleron75 Hey guys.. where we at on this? will this be pushed to the next release?

this PR seems stuck: @p-bystritsky how can we help you?

**Comment by johncarlson21** [comment]
> > @p-bystritsky @aleron75 Hey guys.. where we at on this? will this be pushed to the next release?
> 
> this PR seems stuck: @p-bystritsky how can we help you?

@p-bystritsky @aleron75
guys can we get this moving? I need this for a launch.. or is there a way we can have someone pickup where this left off?

**Comment by p-bystritsky** [comment]
@johncarlson21 @aleron75 I have not so much time, but I'm still working on it.

**Comment by johncarlson21** [comment]
@p-bystritsky  is there anything we can do to help? I really need this fix complete.

**Comment by aleron75** [comment]
This PR adds public methods and changes public methods' parameters, indeed the Semantic Version Checker suggests a MAJOR version change.
That's out of my reach I fear. Let me ask for some help and try to understand how to proceed.

**Comment by johncarlson21** [comment]
Correct me if I'm wrong, but did this really need to change so many files? I mean, there should have been some changes in catalog rules files and option price calculations.
why are there changes to values and other files? 
This problem should have been able to be handled in a couple files, being the problem is in calculation of the product options, and the output of the product option prices

**Comment by p-bystritsky** [comment]
@aleron75 I hope none of my changes are backward incompatible. Anyway, I'll request some more reviews.
@johncarlson21 if you have some suggestions on this PR, you can make your proposals through review.

File: app/code/Magento/CatalogRule/Model/Rule.php [hunk 3]
```diff
@@ -895,4 +898,12 @@ public function getIdentities()
     {
         return ['price'];
     }
+
+    /**
+     * Clear price rules cache.
+     */
+    public function clearPriceRulesData()
```
Comment (sivaschenko) [inline app/code/Magento/CatalogRule/Model/Rule.php]:
```suggestion
    public function clearPriceRulesData(): void
```

File: app/code/Magento/Catalog/Model/Product/Option.php [hunk 4]
```diff
@@ -439,10 +448,13 @@ public function afterSave()
      */
     public function getPrice($flag = false)
     {
-        if ($flag && $this->getPriceType() == self::$typePercent) {
-            $basePrice = $this->getProduct()->getPriceInfo()->getPrice(BasePrice::PRICE_CODE)->getValue();
-            $price = $basePrice * ($this->_getData(self::KEY_PRICE) / 100);
-            return $price;
+        if ($flag) {
+            $product = $this->getProduct();
```
Comment (sivaschenko) [inline app/code/Magento/Catalog/Model/Product/Option.php]: I think `$prorduct` variable is unnecessary

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the PR @p-bystritsky . Please see my review comments

**Comment by engcom-Foxtrot** [comment]
@sivaschenko updated.

**Comment by johncarlson21** [comment]
how is this coming? 
v2.3.3 has been released without this fix and really I cannot believe that no one else is pushing for this fix. Does no one use catalog price rules?

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-5122](https://jira.corp.magento.com/browse/ENGCOM-5122) has been created to process this Pull Request

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-5122](https://jira.corp.magento.com/browse/ENGCOM-5122) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Comment by p-bystritsky** [comment]
@magento run all tests

**Comment by sivaschenko** [comment]
@magento run all tests

**Comment by sdzhepa** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by sdzhepa** [comment]
@magento run all tests

**Comment by johncarlson21** [comment]
guys where are we at with this? I cannot believe that no one else is making a stink about this issue.

**Comment by johncarlson21** [comment]
so what release will this be in?

**Comment by kunzi** [comment]
@johncarlson21 I guess it is going to be in 2.4 as it is in 2.4-develop branch. Not sure... I'm having this issue in M 1.9.3.1 too.

**Comment by johncarlson21** [comment]
@kunzi thanks.. I never had this problem in 1.9.x and that is why when I updated I didn't notice it right away, but our site relys on this function, can't believe it was missed.
Do you have any custom code etc that could be causing the problem on your site?

P.S. if anyone needs a quick fix for the above issue before the release, let me know as I've taken this code and put it into an extension for now.

**Comment by kunzi** [comment]
@johncarlson21 We actually do have an extension in M 1.9.3.1 that sets absolute price for each custom option. Maybe that causes it. Thanks for the hint.

**Comment by gsbhati2008** [comment]
> 
> 
> @kunzi thanks.. I never had this problem in 1.9.x and that is why when I updated I didn't notice it right away, but our site relys on this function, can't believe it was missed.
> Do you have any custom code etc that could be causing the problem on your site?
> 
> P.S. if anyone needs a quick fix for the above issue before the release, let me know as I've taken this code and put it into an extension for now.

Can you please send me extension for now as I need this now.

**Pull request merged**